ARCs/arc-0018.md

```diff
@@ -0,0 +1,348 @@
+
+>The Royalty Collector SHOULD be opted into the full set of assets contained in this list but MAY NOT be checked ahead of time. In the case that an account is not opted into an asset, any transfers where payment is specified for that asset will fail until the account opts into the asset.
+
+
+##### Transfer: 
+
+    transfer(royalty_asset: asset, from: account, to: account, royalty_collector: account, royalty_asset_amount: uint64, payment: transaction)
+    transfer(royalty_asset: asset, from: account, to: account, royalty_collector: account, royalty_asset_amount: uint64, payment: transaction, payment_asset: asset)
+
+Transfers the Asset after checking that the royalty policy is adhered to. This call must be sent by the current owner (the `from` parameter)
+
+The `royalty_asset` is the ASA ID to be transferred. It MUST have a policy defined prior to attempting a transfer.
+
+The `from` parameter is the address for the account the ASA is transferred from.
+
+The `to` parameter is the address for the account the ASA is transferred to.
+
+The `royalty_receiver` parameter is the address for the account that collects the royalty and takes care of any redistribution logic (if any). 
+
+The `royalty_asset_amount` parameter is the number of units of this ASA ID to transfer. The amount MUST be less than or equal to the amount `offered` by the `from` account.
+
+The `payment` parameter is a reference to the transaction that is transferring some asset (ASA or Algos) from the buyer to the Application Address. 
+
+The `payment_asset` parameter is used in the case that the payment is being made with some asset rather than with Algos. It MUST match the Asset ID of the AssetTransfer payment transaction. 
+
+The transfer call SHOULD be part of a group with a size of 2 (app call + payment/asset transfer)
+
+> While its possible to enforce the group size limit, it is possible to circumvent the royalty enforcement logic by simply making an Inner Transaction application call with the appropriate parameters and a small payment, then in the same outer group the "real" payment. The counter-party risk remains the same since the inner transaction is atomic with the outers. 
+> Some ways to prevent a small royalty payment and larger payment in a later transaction of the same group might be by using an `allow` list that is checked against the `auth_addr` of the offer call. The `allow` list would be comprised of known and trusted marketplaces that do not attempt to circumvent the royalty policy. The `allow` list may be implicit as well by transferring a specific asset to the `auth_addr` as frozen and on `offer` a the balance must be > 0 to allow the `auth_addr` to be persisted.
+
+Upon each transfer the the Royalty Enforcer SHOULD assert that `AuthAddr == AssetSender` for the Seller (`from` parameter) and `AuthAddr == AssetReceiver` for the Buyer (`to` parameter). In other words Seller and Buyer accounts not have been Rekeyed. Exceptions can be handled with `Move` method.
```

Comment:
I'm not clear exactly what the intention is here. Do you want to ensure that the Seller or Buyer has never been rekeyed? Or just not within the current group?

If it is a requirement that the accounts have never been required this seems unreasonable given that Rekeying is a very commonly used standard feature of Algorand.

Reply:
Right, this a non-required bit of the spec that calls out a possible way to effectively transfer the asset without going through whatever pre-requisites are required by the contract. 

A holder of this asset (or more generally any controlled asset, think of a security) might need to go through some KYC/AML that allows them to hold it. Using the rekey feature in an atomic transaction allows an account to atomically "sell" the asset.

I don't want to make it required here and think most implementers probably don't need to do this